gurmukhiutils/unicode.py

```diff
@@ -0,0 +1,448 @@
+        >>> print(generated_diacritic_order)
+        '਼ਹਰਵਟਤਨਚਿੇੈੋੌੁੂਾੀਁੱਂੰਃ'
+   """
+    generated_diacritic_order = "".join(
+        nukta_udaat_yakash_order_list
+        + [subjoin_constructor]
+        + [subjoin_consonants]
+        + vowel_order_list
+        + remaining_diacritic_order_list
+    )
+
+    def regex_sort_func(match: Match) -> str:
+        """
+        This re-arranges characters in "match" according to a custom sort order "generated_diacritic_order"
+        """
+        if len(match := match.group()) > 1:
+            match = sorted(match, key=lambda e: generated_diacritic_order.index(e))
+            match = "".join(match)
+
+        return match
+
+    string = re.sub(
+        regex_match_pattern,
+        regex_sort_func,
+        string,
+    )
+
+    return string
+
+
+def sanitize_unicode(string: str) -> str:
+    """
+    Use single char representations of constructed characters.
+    """
+
+    unicode_sanitization_map = {
+        "\u0a73\u0a4b": "\u0a13",  # ਓ
+        "\u0a05\u0a3e": "\u0a06",  # ਅ + ਾ = ਆ
+        "\u0a72\u0a3f": "\u0a07",  # ਇ
+        "\u0a72\u0a40": "\u0a08",  # ਈ
+        "\u0a73\u0a41": "\u0a09",  # ਉ
+        "\u0a73\u0a42": "\u0a0a",  # ਊ
+        "\u0a72\u0a47": "\u0a0f",  # ਏ
+        "\u0a05\u0a48": "\u0a10",  # ਐ
+        "\u0a05\u0a4c": "\u0a14",  # ਔ
+        "\u0a32\u0a3c": "\u0a33",  # ਲ਼
+        "\u0a38\u0a3c": "\u0a36",  # ਸ਼
+        "\u0a59\u0a3c": "\u0a59",  # ਖ਼
+        "\u0a5a\u0a3c": "\u0a5a",  # ਗ਼
+        "\u0a5b\u0a3c": "\u0a5b",  # ਜ਼
+        "\u0a5e\u0a3c": "\u0a5e",  # ਫ਼
+        "\u0a71\u0a02": "\u0a01",  # ਁ adak bindi (quite literally never used today or in the Shabad OS Database, only included for parity with the Unicode block)
+    }
+
+    for key, value in unicode_sanitization_map.items():
+        string = string.replace(key, value)
+
+    return string
+
+
+def decode_unicode(string: str) -> list:
+    """
+    Takes a string and returns a list of keys and values of each character and its corresponding code point.
+
+    Arg:
+        string: The string to affect.
+
+    Returns:
+        A list of each character and its corresponding code point.
+
+    Example:
+        >>> decode_unicode("To ਜੀ")
+        [{'T': '0054'}, {'o': '006f'}, {' ': '0020'}, {'ਜ': '0a1c'}, {'ੀ': '0a40'}]
+    """
+    list = []
+    for item in string:
+        list.append({item: format(ord(item), "04x")})
+
+    return list
+
+
+def encode_unicode(strings: list) -> list:
+    """
+    Takes a string and returns its corresponding unicode character.
+
+    Arg:
+        strings: The list containing any strings to encode.
+
+    Returns:
+        A list of any corresponding unicode characters.
+
+    Examples:
+        >>> encode_unicode(["0054"])
+        'T'
+
+        >>> encode_unicode(["0a1c", "0A40"])
+        '['ਜ', 'ੀ']'
+    """
+    encoded_list = []
+    for string in strings:
+        encoded_list.append(chr(int(string, 16)))
```

Comment:
you can use a list comprehension here
`return [chr(int(string, 16)) for string in strings]`